node/src/components/contract_runtime.rs

```diff
@@ -202,362 +298,341 @@ where
-                let engine_state = Arc::clone(&self.engine_state);
-                let metrics = Arc::clone(&self.metrics);
-                async move {
-                    let correlation_id = CorrelationId::new();
-                    let result = task::spawn_blocking(move || {
-                        let start = Instant::now();
-                        let result = engine_state.read_trie(correlation_id, trie_key);
-                        metrics.read_trie.observe(start.elapsed().as_secs_f64());
-                        result
-                    })
-                    .await
-                    .expect("should run");
-                    let result = match result {
-                        Ok(result) => result,
-                        Err(error) => {
-                            error!(?error, "read_trie_request");
-                            None
+
+                ContractRuntimeResult::GetParentResult {
+                    finalized_block,
+                    deploys,
+                    parent,
+                } => {
+                    trace!(parent_found = %parent.is_some(), finalized_height = %finalized_block.height(), "fetched parent");
+                    let parent_summary = parent.map(|(hash, accumulated_seed, state_root_hash)| {
+                        ExecutedBlockSummary {
+                            hash,
+                            state_root_hash,
+                            accumulated_seed,
                         }
-                    };
-                    trace!(?result, "read_trie response");
-                    responder.respond(result).await
+                    });
+                    self.handle_get_parent_result(
+                        effect_builder,
+                        finalized_block,
+                        deploys,
+                        parent_summary,
+                    )
                 }
-                .ignore()
-            }
-            Event::Request(ContractRuntimeRequest::PutTrie { trie, responder }) => {
-                trace!(?trie, "put_trie request");
-                let engine_state = Arc::clone(&self.engine_state);
-                let metrics = Arc::clone(&self.metrics);
-                async move {
-                    let correlation_id = CorrelationId::new();
-                    let result = task::spawn_blocking(move || {
-                        let start = Instant::now();
-                        let result = engine_state
-                            .put_trie_and_find_missing_descendant_trie_keys(correlation_id, &*trie);
-                        metrics.put_trie.observe(start.elapsed().as_secs_f64());
-                        result
-                    })
-                    .await
-                    .expect("should run");
-                    trace!(?result, "put_trie response");
-                    responder.respond(result).await
+
+                ContractRuntimeResult::RunStepResult { mut state, result } => {
```

Comment:
Since we're not trying to access `self.state`, we should be able to inline the content of this where the `RunStepResult` event is sent.